test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller_phases.go

```diff
@@ -140,36 +144,13 @@ func (r *DockerMachinePoolReconciler) reconcileDockerMachines(ctx context.Contex
 	}
 
 	// Step 1:
-	// Create a DockerMachine for each Docker container so we surface the information to the user. Use the same name as the Docker container for the Docker Machine for ease of lookup.
-	// Providers should iterate through their infrastructure instances and ensure that each instance has a corresponding InfraMachine.
-	for _, machine := range externalMachines {
-		if existingMachine, ok := dockerMachineMap[machine.Name()]; ok {
-			log.V(2).Info("Patching existing DockerMachine", "DockerMachine", klog.KObj(&existingMachine))
```

Comment:
Are we losing this entire branch?

I don't follow how this change solves the problem

Reply:
I moved the creation of the DockerMachine with the Container creation [here](https://github.com/kubernetes-sigs/cluster-api/pull/11440/files#diff-d276dd7b87c738eea06701ab1648df648805fba137c0f4c7bc34c62ba0d55dccR72) to avoid the loop for creating DockerMachine CR based on the previously created container.
 Based on my investigation ex logs blow, for some reason while the DockerMachinePool cleaning all dockerMachine a delay happen for deleting the container which make the DockerMachine get re-created [here](https://github.com/kubernetes-sigs/cluster-api/blob/main/test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller_phases.go#L155) at same time the DockerMachinePool is been deleted
 
 logs ex; the dockerMachine is failed to patch as it is deleted , next log is the dockerMachine created (I think dockerMachinePool go to delete same time the patch is happening to create missing dockerMachine)
 ```
I1119 02:08:36.160435       1 dockermachinepool_controller_phases.go:147] "Patching existing DockerMachine" controller="dockermachinepool" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="DockerMachinePool" DockerMachinePool="self-hosted-jqke01/self-hosted-r77ad9-mp-0-8xsxc" namespace="self-hosted-jqke01" name="self-hosted-r77ad9-mp-0-8xsxc" reconcileID="904d6ace-3baa-48e1-b000-97ed711f9802" MachinePool="self-hosted-r77ad9-mp-0-s8xbz" Cluster="self-hosted-jqke01/self-hosted-r77ad9" DockerMachine="self-hosted-jqke01/worker-1tpa4a"
E1119 02:08:36.181156       1 controller.go:316] "Reconciler error" err="failed to patch DockerMachine self-hosted-jqke01/worker-1tpa4a: dockermachines.infrastructure.cluster.x-k8s.io \"worker-1tpa4a\" not found" controller="dockermachine" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="DockerMachine" DockerMachine="self-hosted-jqke01/worker-1tpa4a" namespace="self-hosted-jqke01" name="worker-1tpa4a" reconcileID="f3f06dcd-f11a-4c3f-ae30-8bbef471ca00"
E1119 02:08:36.183408       1 controller.go:316] "Reconciler error" err="failed to update DockerMachine \"self-hosted-jqke01/worker-1tpa4a\": failed to apply DockerMachine self-hosted-jqke01/worker-1tpa4a: Operation cannot be fulfilled on dockermachines.infrastructure.cluster.x-k8s.io \"worker-1tpa4a\": uid mismatch: the provided object specified uid fb561549-45d7-4a09-af27-6931b56bd020, and no existing object was found" controller="dockermachinepool" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="DockerMachinePool" DockerMachinePool="self-hosted-jqke01/self-hosted-r77ad9-mp-0-8xsxc" namespace="self-hosted-jqke01" name="self-hosted-r77ad9-mp-0-8xsxc" reconcileID="904d6ace-3baa-48e1-b000-97ed711f9802"
I1119 02:08:36.186594       1 dockermachinepool_controller_phases.go:158] "Creating a new DockerMachine for Docker container" controller="dockermachinepool" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="DockerMachinePool" DockerMachinePool="self-hosted-jqke01/self-hosted-r77ad9-mp-0-8xsxc" namespace="self-hosted-jqke01" name="self-hosted-r77ad9-mp-0-8xsxc" reconcileID="9cc3622a-5a30-41c7-8b0a-6f0b212d0bdb" MachinePool="self-hosted-r77ad9-mp-0-s8xbz" Cluster="self-hosted-jqke01/self-hosted-r77ad9" container="worker-1tpa4a"
I1119 02:08:36.253031       1 dockermachine_controller.go:104] "Waiting for Machine Controller to set OwnerRef on DockerMachine" controller="dockermachine" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="DockerMachine" DockerMachine="self-hosted-jqke01/worker-1tpa4a" namespace="self-hosted-jqke01" name="worker-1tpa4a" reconcileID="9799d2ef-c1a7-4924-b347-875155993995" DockerMachinePool="self-hosted-jqke01/self-hosted-r77ad9-mp-0-8xsxc"
```

Reply:
> for some reason while the DockerMachinePool cleaning all dockerMachine a delay happen for deleting the container which make the DockerMachine get re-created here at same time the DockerMachinePool is been deleted

have you considered to simply prevent creation of new machines in this func when the DockerMachinePool has a deletion timestamp? (it should probably be an if around L155-L161)

Reply:
> Well, the code logic shouldn't allow this to happen (it should go through the delete reconcile).

This is confusing me a little bit.
If I got it right, we are trying to fix a race condition that happens when deleting the MP.
But, to fix this error on deletion, we are shuffling machine creation from reconcileDockerMachines to reconcileDockerContainers, and both are not called on reconcile delete 🤔

Reply:
Sorry for confusion, this changes fix the race condition. [here](https://github.com/kubernetes-sigs/cluster-api/blob/main/test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller_phases.go#L156)  during reconcile the dockerMachines if the container has no dockerMachine the dockerMachine get created,   while the delete could happen at sametime.  Before this changes; all containers get created first then all dockerMachines get created (in between race condition could happen). So this changes create every container with its dockerMachine at the sametime to avoid race condition. Hope that make it clear :)

Reply:
Thanks for bearing with me with the explanation.
Ok, we are probably getting to to the point that confuses me.

When I review the PR, I start from the assumption that, if we consider a single MP, controller runtime ensures us that only one concurrent reconcile can happen anytime.

That's means, that if we are running the code [here](https://github.com/kubernetes-sigs/cluster-api/blob/main/test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller_phases.go#L156) we are in reconcile normal, and reconcileDelete won't start until the operation completes or fails, and this rules out most of the race conditions.

Also, If reconcile normal completes, order of operations doesn't matter much, because at the end both dockerMachine and container should exist.

However, what might happen is that it takes some time for the cache to see the newly created dockerMachines, so it will seems it is missing when reconcileDelete starts, even if it it actually exists on the API server (but this requires a different fix).

Does this match with what are you observing? (to check this you can look e.g. at reconcile IDs in logs + look at errors when creating machines)

Reply:
I'm following the patch func [here](https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/util/ssa/patch.go#L64) but not sure how the cache is delayed, Do we run the cache in concurrent exec ?  
But in any case;  attaching the creation of the container with the dockerMachine better implementation than the old way of creating all containers and then create the dockerMachines

Reply:
I'm too new to CAPD for having a strong opinion, but the concern I have is that now creation only happens if _the controller_ creates a Docker container. If one is created in another way, for instance manually, it won't be reconciled into a `DockerMachine`. Not sure if that's a case to consider at all. Just a general slightly bad feeling after seeing hundreds of reconciliation code bugs.

If we're fine about this thread, the rest = LGTM.

Reply:
It's strange to me as well, why do we even create dockerMachines ? as far as I understand we use machinePool to not manage individual machines. If there is a special handling for the machinePool machines then we create machienPoolMachine to perform this special handling (not the case for dockerMachienPool and dockerMachine). 
I'm might be over thinking but this is another discussion over what this PR doing :)

Reply:
Thanks @fabriziopandini for referring to cache implementation. However, what are you referring to is not helping, you can reproduce it by using this repo [here](https://github.com/serngawy/cluster-api/blob/cacheWait/test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller_phases.go#L167)  just build e2e and then run `$ make test-e2e SKIP_RESOURCE_CLEANUP=true GINKGO_FOCUS="self-host"` the test may pass however check the DockerMachine CR in the created kind containers. You will find dockerMachine CR is created  while the DockerMachinePool is deleted and those dockerMachine stuck waiting for its corresponding machine CR get created. 
In any case; what this PR is trying to do is attaching the creation of the container with the dockerMachine which is better implementation from old one to avoid race conditions from happening. 
Can we focus on this PR changes, Is creating the dockerMachine with the container raise an alarm ?

Reply:
Thanks for testing my theory

> You will find dockerMachine CR is created while the DockerMachinePool is deleted and those dockerMachine stuck waiting for its corresponding machine CR get created.

I don't have yet a good explanation about why this happen, but I would prefer to take time to investigate this properly because CR should prevent those race conditions to happen, and if not, we should root cause and get a fix there.

I will try to reproduce locally, but I can't guarantee when I will get to it with the upcoming release and usual EOY rush.

cc @mboersma @AndiDog who might be interested in work this area as well